File: dc_frontend/src/api/resourcePackage.ts
/**
 * 资源包管理API接口
 */

import { request, post, get, put, del } from '@/utils/request'
import type { ApiResponse } from '@/types/datasource'

// 资源包类型枚举
export enum PackageType {
  SQL = 'sql',
  ELASTICSEARCH = 'elasticsearch'
}

// 权限类型枚举
export enum PermissionType {
  READ = 'read',
  WRITE = 'write',
  ADMIN = 'admin'
}

// 条件操作符枚举
export enum ConditionOperator {
  EQ = '=',
  NE = '!=',
  GT = '>',
  GTE = '>=',
  LT = '<',
  LTE = '<=',
  LIKE = 'LIKE',
  IN = 'IN',
  NOT_IN = 'NOT IN',
  IS_NULL = 'IS NULL',
  IS_NOT_NULL = 'IS NOT NULL',
  BETWEEN = 'BETWEEN',
  NOT_BETWEEN = 'NOT BETWEEN'
}

// 逻辑操作符枚举
export enum LogicOperator {
  AND = 'AND',
  OR = 'OR'
}

// 排序方向枚举
export enum SortDirection {
  ASC = 'ASC',
  DESC = 'DESC'
}

// 基础配置接口
export interface BaseConfig {
  schema?: string
  table?: string
  fields: string[]
}

// 条件配置接口
export interface ConditionConfig {
  field: string
  operator: ConditionOperator
  value?: any
  logic: LogicOperator
  description?: string
}

// 动态条件配置接口
export interface DynamicConditionConfig extends ConditionConfig {
  default_value?: any
  required: boolean
  param_name?: string
  param_type?: string
  validation_rules?: Record<string, any>
}

// 排序配置接口
export interface OrderConfig {
  field: string
  direction: SortDirection
}

// 模板接口
export interface Template {
  id: number
  name: string
  description?: string
  type: PackageType
  datasource_id: number
  content: string
  default_params?: Record<string, any>
  is_active: boolean
  created_by: number
  created_at: string
  updated_at: string
}

// 资源包标签接口
export interface ResourcePackageTag {
  id: number
  package_id: number
  tag_name: string
  tag_color: string
  created_at: string
}

// 资源包权限接口
export interface ResourcePackagePermission {
  id: number
  package_id: number
  user_id: number
  permission_type: PermissionType
  granted_by: number
  granted_at: string
  expires_at?: string
  is_active: boolean
}

// 资源包接口
export interface ResourcePackage {
  id: number
  name: string
  description?: string
  type: PackageType
  datasource_id: number
  resource_id?: number
  template_id: number
  template_type: PackageType
  dynamic_params?: Record<string, any>
  is_active: boolean
  is_lock: string
  created_by: number
  created_at: string
  updated_at: string
  // 新增：生成与下载时间/路径
  download_time?: string
  download_url?: string
  excel_time?: string
  tags: ResourcePackageTag[]
  permissions: ResourcePackagePermission[]
}

// 历史文件接口
export interface ResourcePackageFile {
  id: number
  filename: string
  object_path: string
  generated_at: string
}

// 创建资源包请求接口
export interface ResourcePackageCreateRequest {
  name: string
  description?: string
  type: PackageType
  datasource_id: number
  resource_id?: number
  template_id: number
  template_type: PackageType
  dynamic_params?: Record<string, any>
  is_active: boolean
  is_lock?: string
  tags?: string[]
}

// 更新资源包请求接口
export interface ResourcePackageUpdateRequest {
  name?: string
  description?: string
  template_id?: number
  template_type?: PackageType
  dynamic_params?: Record<string, any>
  is_active?: boolean
  is_lock?: string
  tags?: string[]
}

// 搜索请求接口
export interface ResourcePackageSearchRequest {
  keyword?: string
  type?: PackageType
  datasource_id?: number
  created_by?: number
  tags?: string[]
  is_active?: boolean
  is_lock?: string
  page: number
  size: number
  sort_by: string
  sort_order: string
}

// 列表响应接口
export interface ResourcePackageListResponse {
  items: ResourcePackage[]
  total: number
  page: number
  size: number
  pages: number
}

// 查询请求接口
export interface ResourcePackageQueryRequest {
  dynamic_params: Record<string, any>
  limit?: number
  offset: number
  format?: string
}

// 查询响应接口
export interface ResourcePackageQueryResponse {
  success: boolean
  columns: string[]
  data: any[][]
  total_count: number
  execution_time: number
  query_id?: string
  generated_query?: string
  error_message?: string
}

// 参数信息接口
export interface ResourcePackageParamInfo {
  param_name: string
  field: string
  operator: ConditionOperator
  param_type: string
  default_value?: any
  required: boolean
  description?: string
  validation_rules?: Record<string, any>
}



/**
 * 资源包管理API
 */
export const resourcePackageApi = {
  /**
   * 创建资源包
   */
  create(data: ResourcePackageCreateRequest): Promise<ApiResponse<ResourcePackage>> {
    return post('/resource-packages/', data)
  },

  /**
   * 获取资源包详情
   */
  get(id: number): Promise<ApiResponse<ResourcePackage>> {
    return get(`/resource-packages/${id}`)
  },

  /**
   * 更新资源包
   */
  update(id: number, data: ResourcePackageUpdateRequest): Promise<ApiResponse<ResourcePackage>> {
    return put(`/resource-packages/${id}`, data)
  },

  /**
   * 删除资源包
   */
  delete(id: number): Promise<ApiResponse<{ message: string }>> {
    return del(`/resource-packages/${id}`)
  },

  /**
   * 搜索资源包
   */
  search(params: ResourcePackageSearchRequest): Promise<ApiResponse<ResourcePackageListResponse>> {
    return post('/resource-packages/search', params)
  },



  /**
   * 查询资源包数据
   */
  query(id: number, params: ResourcePackageQueryRequest): Promise<ApiResponse<ResourcePackageQueryResponse>> {
    return post(`/resource-packages/${id}/query`, params)
  },

  /**
   * 生成Excel（更新excel_time与download_url）
   */
  generateExcel(id: number, payload: Record<string, any>): Promise<ApiResponse<any>> {
    // 为长耗时任务单独提升超时到180秒
    return post(`/resource-packages/${id}/generate-excel`, payload, { timeout: 180000 })
  },

  /**
   * 下载最新资源包（更新download_time）
   */
  downloadLatest(id: number): Promise<ApiResponse<any>> {
    return post(`/resource-packages/${id}/download-latest`, {})
  },

  /**
   * 安全查询资源包数据 - 增强安全性版本
   * 专为ResourcePackageQueryPage.vue页面设计
   */
  secureQuery(id: number, params: ResourcePackageQueryRequest): Promise<ApiResponse<ResourcePackageQueryResponse>> {
    return post(`/resource-packages/${id}/secure-query`, params)
  },

  /**
   * 获取安全查询历史记录
   */
  getSecureQueryHistory(id: number, page: number = 1, size: number = 20): Promise<ApiResponse<any>> {
    return get(`/resource-packages/${id}/query-history?page=${page}&size=${size}`)
  },

  /**
   * 获取查询历史
   */
  getHistory(id: number, page: number = 1, size: number = 20): Promise<ApiResponse<any>> {
    return get(`/resource-packages/${id}/history`, {
      page, size
    })
  },

  /**
   * 列出历史Excel文件
   */
  listFiles(id: number, page: number = 1, size: number = 20): Promise<ApiResponse<{ items: ResourcePackageFile[]; total: number; page: number; size: number }>> {
    return get(`/resource-packages/${id}/files`, { page, size })
  },

  /**
   * 下载指定历史文件（返回预签名URL）
   */
  downloadFile(id: number, fileId: number): Promise<ApiResponse<{ download_url: string; filename: string; object_path: string }>> {
    return post(`/resource-packages/${id}/files/${fileId}/download`, {})
  }
}

/**
 * 模板管理API
 */
export const templateApi = {
  /**
   * 获取模板列表
   */
  list(params?: { datasource_id?: number; type?: PackageType; data_resource_id?: number; indices?: string[] }): Promise<{ data: Template[] }> {
    const query: any = {}
    if (params?.datasource_id) query.datasource_id = params.datasource_id
    if (params?.data_resource_id) query.data_resource_id = params.data_resource_id
    if (params?.indices && params.indices.length > 0) query.indices = params.indices.join(',')

    if (params?.type === PackageType.ELASTICSEARCH) {
      // ES模板端点
      return get('/es/templates', query)
    }
    // 默认SQL模板端点（并标记为模板）
    query.isTemplate = true
    return get('/sql/templates', query)
  },

  /**
   * 获取模板详情
   */
  get(id: number, type: PackageType): Promise<{ data: Template }> {
    if (type === PackageType.ELASTICSEARCH) {
      return get(`/es/templates/${id}`)
    }
    return get(`/sql/templates/${id}`)
  },

  /**
   * 创建模板
   */
  create(data: Omit<Template, 'id' | 'created_at' | 'updated_at' | 'created_by'>): Promise<Template> {
    return post('/templates/', data)
  },

  /**
   * 更新模板
   */
  update(id: number, data: Partial<Omit<Template, 'id' | 'created_at' | 'updated_at' | 'created_by'>>): Promise<Template> {
    return put(`/templates/${id}`, data)
  },

  /**
   * 删除模板
   */
  delete(id: number): Promise<{ message: string }> {
    return del(`/templates/${id}`)
  }
}

export default resourcePackageApi

File: backend/app/schemas/resource_package.py
"""资源包相关的Pydantic模式定义"""

from datetime import datetime
from typing import Optional, List, Dict, Any, Union
from pydantic import BaseModel, Field, validator
from enum import Enum


class PackageType(str, Enum):
    """资源包类型枚举"""
    SQL = "sql"
    ELASTICSEARCH = "elasticsearch"


class PermissionType(str, Enum):
    """权限类型枚举"""
    READ = "read"
    WRITE = "write"
    ADMIN = "admin"


class QueryStatus(str, Enum):
    """查询状态枚举"""
    SUCCESS = "success"
    ERROR = "error"
    TIMEOUT = "timeout"


class ConditionOperator(str, Enum):
    """条件操作符枚举"""
    EQ = "="
    NE = "!="
    GT = ">"
    GTE = ">="
    LT = "<"
    LTE = "<="
    LIKE = "LIKE"
    IN = "IN"
    NOT_IN = "NOT IN"
    IS_NULL = "IS NULL"
    IS_NOT_NULL = "IS NOT NULL"
    BETWEEN = "BETWEEN"
    NOT_BETWEEN = "NOT BETWEEN"


class LogicOperator(str, Enum):
    """逻辑操作符枚举"""
    AND = "AND"
    OR = "OR"


class SortDirection(str, Enum):
    """排序方向枚举"""
    ASC = "ASC"
    DESC = "DESC"


# 基础配置模式
class BaseConfig(BaseModel):
    """基础配置模式"""
    schema: Optional[str] = Field(None, description="数据库/索引名称")
    table: Optional[str] = Field(None, description="表名/索引名")
    fields: List[str] = Field(default_factory=list, description="字段列表")
    
    class Config:
        extra = "allow"


# 条件配置模式
class ConditionConfig(BaseModel):
    """条件配置模式"""
    field: str = Field(..., description="字段名")
    operator: ConditionOperator = Field(..., description="操作符")
    value: Optional[Union[str, int, float, List[Any]]] = Field(None, description="值")
    logic: LogicOperator = Field(LogicOperator.AND, description="逻辑操作符")
    description: Optional[str] = Field(None, description="条件描述")


# 动态条件配置模式
class DynamicConditionConfig(ConditionConfig):
    """动态条件配置模式"""
    default_value: Optional[Union[str, int, float, List[Any]]] = Field(None, description="默认值")
    required: bool = Field(False, description="是否必填")
    param_name: Optional[str] = Field(None, description="参数名称")
    param_type: Optional[str] = Field("string", description="参数类型")
    validation_rules: Optional[Dict[str, Any]] = Field(None, description="验证规则")


# 排序配置模式
class OrderConfig(BaseModel):
    """排序配置模式"""
    field: str = Field(..., description="排序字段")
    direction: SortDirection = Field(SortDirection.ASC, description="排序方向")


# 标签模式
class ResourcePackageTagBase(BaseModel):
    """资源包标签基础模式"""
    tag_name: str = Field(..., max_length=100, description="标签名称")
    tag_color: str = Field("#409EFF", max_length=20, description="标签颜色")


class ResourcePackageTagCreate(ResourcePackageTagBase):
    """创建资源包标签模式"""
    pass


class ResourcePackageTag(ResourcePackageTagBase):
    """资源包标签模式"""
    id: int
    package_id: int
    created_at: datetime
    
    class Config:
        from_attributes = True


# 权限模式
class ResourcePackagePermissionBase(BaseModel):
    """资源包权限基础模式"""
    user_id: int = Field(..., description="用户ID")
    permission_type: PermissionType = Field(..., description="权限类型")
    expires_at: Optional[datetime] = Field(None, description="过期时间")


class ResourcePackagePermissionCreate(ResourcePackagePermissionBase):
    """创建资源包权限模式"""
    pass


class ResourcePackagePermission(ResourcePackagePermissionBase):
    """资源包权限模式"""
    id: int
    package_id: int
    granted_by: int
    granted_at: datetime
    is_active: bool
    
    class Config:
        from_attributes = True


# 查询历史模式
class ResourcePackageQueryHistoryBase(BaseModel):
    """资源包查询历史基础模式"""
    dynamic_params: Optional[Dict[str, Any]] = Field(None, description="动态参数值")
    generated_query: Optional[str] = Field(None, description="生成的查询语句")
    result_count: int = Field(0, description="结果行数")
    execution_time: int = Field(0, description="执行时间(毫秒)")
    status: QueryStatus = Field(QueryStatus.SUCCESS, description="执行状态")
    error_message: Optional[str] = Field(None, description="错误信息")


class ResourcePackageQueryHistory(ResourcePackageQueryHistoryBase):
    """资源包查询历史模式"""
    id: int
    package_id: int
    user_id: int
    created_at: datetime
    
    class Config:
        from_attributes = True


# 资源包模式
class ResourcePackageBase(BaseModel):
    """资源包基础模式"""
    name: str = Field(..., max_length=255, description="资源包名称")
    description: Optional[str] = Field(None, description="资源包描述")
    type: PackageType = Field(..., description="资源包类型")
    datasource_id: int = Field(..., description="数据源ID")
    resource_id: Optional[int] = Field(None, description="数据资源ID")
    template_id: Optional[int] = Field(None, description="查询模板ID")
    template_type: str = Field(..., description="模板类型(sql/elasticsearch)")
    dynamic_params: Optional[Dict[str, Any]] = Field(None, description="动态参数")
    is_active: bool = Field(True, description="是否启用")


class ResourcePackageCreate(ResourcePackageBase):
    """创建资源包模式"""
    tags: Optional[List[str]] = Field(default_factory=list, description="标签列表")
    
    @validator('name')
    def validate_name(cls, v):
        if not v or not v.strip():
            raise ValueError('资源包名称不能为空')
        return v.strip()
    
    @validator('template_type')
    def validate_template_type(cls, v):
        if v not in ['sql', 'elasticsearch']:
            raise ValueError('模板类型必须是sql或elasticsearch')
        return v


class ResourcePackageUpdate(BaseModel):
    """更新资源包模式"""
    name: Optional[str] = Field(None, max_length=255, description="资源包名称")
    description: Optional[str] = Field(None, description="资源包描述")
    template_id: Optional[int] = Field(None, description="查询模板ID")
    template_type: Optional[str] = Field(None, description="模板类型(sql/elasticsearch)")
    dynamic_params: Optional[Dict[str, Any]] = Field(None, description="动态参数")
    is_active: Optional[bool] = Field(None, description="是否启用")
    tags: Optional[List[str]] = Field(None, description="标签列表")


class ResourcePackage(ResourcePackageBase):
    """资源包模式"""
    id: int
    created_by: int
    created_at: datetime
    updated_at: datetime
    tags: List[ResourcePackageTag] = Field(default_factory=list, description="标签列表")
    permissions: List[ResourcePackagePermission] = Field(default_factory=list, description="权限列表")
    
    class Config:
        from_attributes = True


class ResourcePackageResponse(BaseModel):
    """资源包响应模式（不包含dynamic_params）"""
    id: int
    name: str = Field(..., max_length=255, description="资源包名称")
    description: Optional[str] = Field(None, description="资源包描述")
    type: PackageType = Field(..., description="资源包类型")
    datasource_id: int = Field(..., description="数据源ID")
    resource_id: Optional[int] = Field(None, description="数据资源ID")
    template_id: Optional[int] = Field(None, description="查询模板ID")
    template_type: str = Field(..., description="模板类型(sql/elasticsearch)")
    is_active: bool = Field(True, description="是否启用")
    created_by: int
    created_at: datetime
    updated_at: datetime
    tags: List[ResourcePackageTag] = Field(default_factory=list, description="标签列表")
    permissions: List[ResourcePackagePermission] = Field(default_factory=list, description="权限列表")
    
    class Config:
        from_attributes = True


# 查询请求模式
class ResourcePackageQueryRequest(BaseModel):
    """资源包查询请求模式"""
    dynamic_params: Dict[str, Any] = Field(default_factory=dict, description="动态参数值")
    limit: Optional[int] = Field(None, ge=1, le=10000, description="限制条数")
    offset: int = Field(0, ge=0, description="偏移量")
    format: Optional[str] = Field("json", description="返回格式")


# 查询响应模式
class ResourcePackageQueryResponse(BaseModel):
    """资源包查询响应模式"""
    success: bool = Field(..., description="是否成功")
    columns: List[str] = Field(default_factory=list, description="列名列表")
    data: List[List[Any]] = Field(default_factory=list, description="数据行")
    total_count: int = Field(0, description="总行数")
    execution_time: int = Field(0, description="执行时间(毫秒)")
    query_id: Optional[str] = Field(None, description="查询ID")
    generated_query: Optional[str] = Field(None, description="生成的查询语句")
    error_message: Optional[str] = Field(None, description="错误信息")


# 搜索请求模式
class ResourcePackageSearchRequest(BaseModel):
    """资源包搜索请求模式"""
    keyword: Optional[str] = Field(None, description="关键词")
    type: Optional[PackageType] = Field(None, description="资源包类型")
    datasource_id: Optional[int] = Field(None, description="数据源ID")
    created_by: Optional[int] = Field(None, description="创建者ID")
    tags: Optional[List[str]] = Field(None, description="标签列表")
    is_active: Optional[bool] = Field(None, description="是否启用")
    page: int = Field(1, ge=1, description="页码")
    size: int = Field(20, ge=1, le=100, description="每页大小")
    sort_by: str = Field("created_at", description="排序字段")
    sort_order: str = Field("desc", description="排序方向")


# 列表响应模式
class ResourcePackageListResponse(BaseModel):
    """资源包列表响应模式"""
    items: List[ResourcePackage] = Field(default_factory=list, description="资源包列表")
    total: int = Field(0, description="总数")
    page: int = Field(1, description="当前页")
    size: int = Field(20, description="每页大小")
    pages: int = Field(0, description="总页数")


# 参数信息模式
class ResourcePackageParamInfo(BaseModel):
    """资源包参数信息模式"""
    param_name: str = Field(..., description="参数名称")
    field: str = Field(..., description="字段名")
    operator: ConditionOperator = Field(..., description="操作符")
    param_type: str = Field("string", description="参数类型")
    default_value: Optional[Any] = Field(None, description="默认值")
    required: bool = Field(False, description="是否必填")
    description: Optional[str] = Field(None, description="参数描述")
    validation_rules: Optional[Dict[str, Any]] = Field(None, description="验证规则")